code/game/machinery/computer/communications.dm

```diff
@@ -70,85 +72,76 @@
 			return
-
+		// Login function.
 		var/list/access = usr.get_access()
 		if(allowed(usr))
 			authenticated = COMM_AUTHENTICATION_MIN
-
 		if(ACCESS_CAPTAIN in access)
 			authenticated = COMM_AUTHENTICATION_MAX
 			var/mob/living/carbon/human/H = usr
 			var/obj/item/card/id = H.get_idcard(TRUE)
 			if(istype(id))
 				crew_announcement.announcer = GetNameAndAssignmentFromId(id)
-
-		SSnanoui.update_uis(src)
-		return
-
-	if(href_list["logout"])
-		authenticated = COMM_AUTHENTICATION_NONE
-		crew_announcement.announcer = ""
-		setMenuState(usr,COMM_SCREEN_MAIN)
-		SSnanoui.update_uis(src)
+		if(authenticated == COMM_AUTHENTICATION_NONE)
+			to_chat(usr, "<span class='warning'>You need to swipe your ID.</span>")
 		return
 
+	// All functions below this point require authentication.
 	if(!is_authenticated(usr))
-		return 1
+		return FALSE
 
-	switch(href_list["operation"])
+	switch(action)
 		if("main")
-			setMenuState(usr,COMM_SCREEN_MAIN)
-
-		if("changeseclevel")
-			setMenuState(usr,COMM_SCREEN_SECLEVEL)
+			setMenuState(usr, COMM_SCREEN_MAIN)
 
 		if("newalertlevel")
 			if(isAI(usr) || isrobot(usr))
 				to_chat(usr, "<span class='warning'>Firewalls prevent you from changing the alert level.</span>")
-				return 1
+				return
 			else if(usr.can_admin_interact())
-				change_security_level(text2num(href_list["level"]))
-				return 1
+				change_security_level(text2num(params["level"]))
+				return
 			else if(!ishuman(usr))
 				to_chat(usr, "<span class='warning'>Security measures prevent you from changing the alert level.</span>")
-				return 1
-
-			var/mob/living/carbon/human/L = usr
-			var/obj/item/card = L.get_active_hand()
-			var/obj/item/card/id/I = (card && card.GetID()) || L.wear_id || L.wear_pda
-			if(istype(I, /obj/item/pda))
-				var/obj/item/pda/pda = I
-				I = pda.id
-			if(I && istype(I))
+				return
+
+			var/mob/living/carbon/human/H = usr
+			var/obj/item/card/id/I = H.get_idcard(TRUE)
+			if(istype(I))
 				if(ACCESS_CAPTAIN in I.access)
-					change_security_level(text2num(href_list["level"]))
+					change_security_level(text2num(params["level"]))
 				else
 					to_chat(usr, "<span class='warning'>You are not authorized to do this.</span>")
-				setMenuState(usr,COMM_SCREEN_MAIN)
+				setMenuState(usr, COMM_SCREEN_MAIN)
 			else
 				to_chat(usr, "<span class='warning'>You need to swipe your ID.</span>")
 
 		if("announce")
 			if(is_authenticated(usr) == COMM_AUTHENTICATION_MAX)
 				if(message_cooldown)
 					to_chat(usr, "<span class='warning'>Please allow at least one minute to pass between announcements.</span>")
-					SSnanoui.update_uis(src)
 					return
 				var/input = input(usr, "Please write a message to announce to the station crew.", "Priority Announcement")
 				if(!input || message_cooldown || ..() || !(is_authenticated(usr) == COMM_AUTHENTICATION_MAX))
-					SSnanoui.update_uis(src)
 					return
 				crew_announcement.Announce(input)
 				message_cooldown = 1
```

Comment:
Any change you can take this and the spawn below it to use booleans, as well as `addtimer` instead of the `spawn()`